src/main/resources/liquibase/crop_changelog/23_1_0.xml

```diff
@@ -64,5 +64,339 @@
+            <and>
+                <not>
+                    <sqlCheck expectedResult="0">
+                        SELECT COUNT(*) FROM names name
+                        LEFT JOIN udflds nametype ON nametype.fldno = name.ntype
+                        WHERE nametype.fldno IS NULL;
+                    </sqlCheck>
+                </not>
+                <not>
+                    <tableExists tableName="names_invalid_ntype_bkp"/>
+                </not>
+            </and>
+        </preConditions>
+        <comment> Create names_invalid_ntype_bkp table</comment>
+        <sql dbms="mysql" splitStatements="true">
+            CREATE TABLE names_invalid_ntype_bkp AS
+            SELECT name.* FROM names name
+            LEFT JOIN udflds nametype ON nametype.fldno = name.ntype
+            WHERE nametype.fldno IS NULL;
+
+            ALTER TABLE names_invalid_ntype_bkp ADD changed_to_lname TINYINT(1) DEFAULT 0 NOT NULL;
+        </sql>
+    </changeSet>
+
+    <changeSet author="mderamos" id="v23.1.0-3">
+        <preConditions onFail="MARK_RAN">
+            <and>
+                <sqlCheck expectedResult="0">
+                    SELECT COUNT(*) FROM information_schema.routines
+                    WHERE routine_schema = DATABASE() AND specific_name = 'fix_names_with_invalid_name_type';
+                </sqlCheck>
+                <tableExists tableName="names_invalid_ntype_bkp"/>
+            </and>
+        </preConditions>
+        <comment>
+            Create stored procedure to "fix" names with invalid name types. Names with invalid nametype will be deleted
+            except for cases where all nametypes for a germplasm is invalid. The deleted name entries will be added
+            to a names backup table.
+        </comment>
+        <sql dbms="mysql" splitStatements="false">
+            <![CDATA[
+            CREATE PROCEDURE `fix_names_with_invalid_name_type`()
+            BEGIN
+                DECLARE gid INT;
+                DECLARE isPreferredNameInvalid INT DEFAULT false;
+                DECLARE hasValidNames INT DEFAULT false;
+                DECLARE hasValidPreferredName INT DEFAULT false;
+                DECLARE hasInvalidPreferredName INT DEFAULT false;
+                DECLARE done INT DEFAULT false;
+                DECLARE validNameId INT;
+                DECLARE invalidNameId INT;
+
+
+                -- retrieve germplasm with invalid preferred name types
+                DECLARE gid_with_invalid_name_types CURSOR FOR
+                    SELECT DISTINCT(name.gid) as `gid` FROM names_invalid_ntype_bkp name;
+
+                DECLARE CONTINUE HANDLER FOR NOT FOUND SET done = TRUE;
+
+                DECLARE EXIT HANDLER FOR SQLEXCEPTION
+                BEGIN
+                    ROLLBACK;
+                    SIGNAL SQLSTATE '45000' SET MESSAGE_TEXT = 'Error in fixing names with invalid nameTypes.';
+                END;
 
+                OPEN gid_with_invalid_name_types;
```

Comment:
@mderamos I am concerned about opening a cursor and iterating case by case, there should be way to simplify it. Not only making it more readable but improving the performance too, like creating a temporary table and identify the different scenarios. 
When writing migrations we should avois as much as possible to iterate when we can run massive update or delete sentences. 
I was doing some analysis on this first SP, I will share with you my ideas, I did not have time to test it and complete it but I want to share it and see if you went into a similar exercise. 
CC @abatac